File: frontend-design/scripts/search_frontend_design.py
#!/usr/bin/env python3
"""Search frontend-design references and CSV data with stdlib only.

Usage:
  python scripts/search_frontend_design.py dashboard
  python scripts/search_frontend_design.py "trust finance" --data-only
  python scripts/search_frontend_design.py motion --refs-only -n 8
"""

from __future__ import annotations

import argparse
import csv
import re
from pathlib import Path
from math import log
from _catalog import expand_query

ROOT = Path(__file__).resolve().parents[1]
REFS = ROOT / "references"
DATA = ROOT / "data"


def tokenize(text: str) -> list[str]:
    return re.findall(r"[a-zA-Z0-9_+-]+", text.lower())


def score(query: list[str], text: str) -> float:
    tokens = tokenize(text)
    if not tokens:
        return 0.0
    counts: dict[str, int] = {}
    for token in tokens:
        counts[token] = counts.get(token, 0) + 1
    value = 0.0
    length_norm = 1 + log(len(tokens) + 1)
    for term in query:
        value += counts.get(term, 0) / length_norm
        for token in counts:
            if term in token and term != token:
                value += 0.25 / length_norm
    return value


def iter_reference_chunks() -> list[dict[str, str]]:
    chunks: list[dict[str, str]] = []
    for path in sorted(REFS.glob("*.md")):
        text = path.read_text(encoding="utf-8")
        current_title = path.name
        current: list[str] = []
        for line in text.splitlines():
            if line.startswith("## ") and current:
                chunks.append({"source": str(path.relative_to(ROOT)), "title": current_title, "text": "\n".join(current).strip()})
                current = []
            if line.startswith("#"):
                current_title = line.lstrip("# ").strip() or current_title
            current.append(line)
        if current:
            chunks.append({"source": str(path.relative_to(ROOT)), "title": current_title, "text": "\n".join(current).strip()})
    return chunks


def format_row(row: dict[str, object]) -> str:
    parts: list[str] = []
    for key, value in row.items():
        if value is None:
            continue
        if isinstance(value, list):
            rendered = ", ".join(str(item) for item in value)
        else:
            rendered = str(value)
        parts.append(f"{key}: {rendered}")
    return " | ".join(parts)


def iter_data_rows() -> list[dict[str, str]]:
    rows: list[dict[str, str]] = []
    for path in sorted(DATA.glob("*.csv")):
        with path.open(newline="", encoding="utf-8") as handle:
            for index, row in enumerate(csv.DictReader(handle), start=1):
                title = str(row.get("canonical") or row.get("style") or row.get("product_category") or row.get("Product Type") or row.get("Style Category") or row.get("Pattern Name") or f"row {index}")
                rows.append({"source": str(path.relative_to(ROOT)), "title": title, "text": format_row(row)})
    return rows


def main() -> int:
    parser = argparse.ArgumentParser(description="Search frontend-design skill references")
    parser.add_argument("query", nargs="+", help="Search terms")
    parser.add_argument("-n", "--limit", type=int, default=6)
    parser.add_argument("--refs-only", action="store_true")
    parser.add_argument("--data-only", action="store_true")
    args = parser.parse_args()
    query = expand_query(" ".join(args.query))
    docs: list[dict[str, str]] = []
    if not args.data_only:
        docs.extend(iter_reference_chunks())
    if not args.refs_only:
        docs.extend(iter_data_rows())
    ranked = sorted(((score(query, doc["text"] + " " + doc["title"]), doc) for doc in docs), key=lambda item: item[0], reverse=True)
    shown = 0
    for value, doc in ranked:
        if value <= 0:
            continue
        shown += 1
        excerpt = re.sub(r"\s+", " ", doc["text"]).strip()
        if len(excerpt) > 520:
            excerpt = excerpt[:517] + "..."
        print(f"## {doc['title']}\nSource: {doc['source']} | score={value:.3f}\n{excerpt}\n")
        if shown >= args.limit:
            break
    if shown == 0:
        print("No matches")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: frontend-design/scripts/search_frontend_design.mjs
#!/usr/bin/env node
/** Search frontend-design references and CSV data with dependency-free Node ESM.
 *
 * Usage:
 *   node scripts/search_frontend_design.mjs dashboard
 *   node scripts/search_frontend_design.mjs "trust finance" --data-only
 *   node scripts/search_frontend_design.mjs motion --refs-only -n 8
 */

import { readdirSync, readFileSync } from "node:fs";
import { basename, relative, resolve } from "node:path";
import { pathToFileURL } from "node:url";
import { DATA, REFS, expandQuery, loadCsv, tokenize } from "./_catalog.mjs";

const ROOT = resolve(DATA, "..");

function parseArgs(args) {
  let limit = 6;
  let refsOnly = false;
  let dataOnly = false;
  const query = [];
  for (let index = 0; index < args.length; index += 1) {
    const argument = args[index];
    if (argument === "-n" || argument === "--limit") {
      limit = Number.parseInt(args[index + 1], 10);
      index += 1;
    } else if (argument === "--refs-only") {
      refsOnly = true;
    } else if (argument === "--data-only") {
      dataOnly = true;
    } else {
      query.push(argument);
    }
  }
  if (query.length === 0 || !Number.isInteger(limit)) throw new Error("Usage: search_frontend_design.mjs <query> [-n <limit>] [--refs-only|--data-only]");
  return { query: query.join(" "), limit, refsOnly, dataOnly };
}

function searchScore(query, text) {
  const tokens = tokenize(text);
  if (tokens.length === 0) return 0;
  const counts = new Map();
  for (const token of tokens) counts.set(token, (counts.get(token) ?? 0) + 1);
  const norm = 1 + Math.log(tokens.length + 1);
  let value = 0;
  for (const term of query) {
    value += (counts.get(term) ?? 0) / norm;
    for (const token of counts.keys()) {
      if (term !== token && token.includes(term)) value += 0.25 / norm;
    }
  }
  return value;
}

function referenceChunks() {
  const chunks = [];
  for (const name of readdirSync(REFS).filter((entry) => entry.endsWith(".md")).sort()) {
    const text = readFileSync(resolve(REFS, name), "utf8");
    let currentTitle = name;
    let current = [];
    for (const line of text.split(/\r?\n/)) {
      if (line.startsWith("## ") && current.length > 0) {
        chunks.push({ source: relative(ROOT, resolve(REFS, name)), title: currentTitle, text: current.join("\n").trim() });
        current = [];
      }
      if (line.startsWith("#")) currentTitle = line.replace(/^#+\s*/, "").trim() || currentTitle;
      current.push(line);
    }
    if (current.length > 0) chunks.push({ source: relative(ROOT, resolve(REFS, name)), title: currentTitle, text: current.join("\n").trim() });
  }
  return chunks;
}

function formatRow(row) {
  return Object.entries(row).filter(([, value]) => value != null).map(([key, value]) => `${key}: ${Array.isArray(value) ? value.join(", ") : value}`).join(" | ");
}

function dataRows() {
  const rows = [];
  for (const name of readdirSync(DATA).filter((entry) => entry.endsWith(".csv")).sort()) {
    loadCsv(name).forEach((row, index) => {
      const title = row.canonical || row.style || row.product_category || row["Product Type"] || row["Style Category"] || row["Pattern Name"] || `row ${index + 1}`;
      rows.push({ source: `data/${basename(name)}`, title, text: formatRow(row) });
    });
  }
  return rows;
}

export function main(args = process.argv.slice(2)) {
  let options;
  try {
    options = parseArgs(args);
  } catch (error) {
    console.error(error.message);
    return 2;
  }
  const query = expandQuery(options.query);
  const documents = [];
  if (!options.dataOnly) documents.push(...referenceChunks());
  if (!options.refsOnly) documents.push(...dataRows());
  const ranked = documents.map((document) => [searchScore(query, `${document.text} ${document.title}`), document]).sort((left, right) => right[0] - left[0]);
  const lines = [];
  let shown = 0;
  for (const [value, document] of ranked) {
    if (value <= 0) continue;
    let excerpt = document.text.replace(/\s+/g, " ").trim();
    if (excerpt.length > 520) excerpt = `${excerpt.slice(0, 517)}...`;
    lines.push(`## ${document.title}`, `Source: ${document.source} | score=${value.toFixed(3)}`, excerpt, "");
    shown += 1;
    if (shown >= options.limit) break;
  }
  if (shown === 0) lines.push("No matches");
  process.stdout.write(`${lines.join("\n")}\n`);
  return 0;
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) process.exitCode = main();
